firefox-ios/Client/Frontend/Home/Homepage Rebuild/TopSites/TopSitesDimensionImplementation.swift

```diff
@@ -2,13 +2,38 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
-struct TopSitesDimensionImplementation {
-    /// Get the number of tiles (top sites) per row the user will see. This depends on the UI interface the user has.
+import Common
+
+class TopSitesDimensionImplementation {
+    /// The update count of number of tiles per row based on device layout
+    /// After updating the value, the top sites state should be updated respectively
+    var numberOfTilesPerRow: Int? {
+        willSet {
+            guard newValue != numberOfTilesPerRow else { return }
+            queue.async {
```

Comment:
I thought that dispatching actions were already only dispatching on main thread, so not sure why I need to add this, but I was getting a crash and ensuring the call happens on main thread, addresses the issue. Going to look at this again with fresh eyes on Monday, but opened a PR to review to get early eyes + if anyone has a clue.

Reply:
Will need to spend more time investigating, I'll open a new ticket to investigate this separately to avoid blocking the other changes for this PR.

Reply:
ya we should definitely not need to do this here. 
```
 public func dispatch(_ action: Action) {
        logger.log("Dispatched action: \(action.displayString())", level: .info, category: .redux)

        guard Thread.isMainThread else {
            DispatchQueue.main.async { [weak self] in self?.dispatch(action) }
            return
        }
```
Should be already dispatching the work on the main thread. I will pull down your branch and see what I can find!

Reply:
@Cramsden It seems the issue is because `updateNumberOfTilesPerRow` gets called pretty frequently and causes multiple dispatches when unnecessary, so the comparison of `newValue` and `numberOfTilesRow` isn't accurate. To avoid confusion, instead I wrapped `queue.async` in when that property gets set. Let me know if you have any concerns or if you find anything else interesting!

Reply:
On a separate, but related note, I do see some flaky tests that failed on other PRs as well related to dispatching multiple actions on a separate threads.

Seems like the first failed instance was on [Nov 12](https://app.bitrise.io/insights/095364a9afb324b2/explore/tests?interval=daily&duration=FIXED&tab=failure_rate&app_slug=6c06d3a40422d10f&test_suite_name=ReduxTests&display_option=top10&secondary_tab=related_test_case_runs&breakdown=cc%2FFXIOS-10913_add-number-of-rows-for-top-site-cal&breakdown=Bugfix-FXIOS-10575-Protection-panel-Clear-Cookies-Site-Data-DesignQA&breakdown=DD%2F10008_update-bookmark-panel-edit-p2&breakdown=DD%2F9766_update-bookmarks-panel-edit-mode-p1&breakdown=Refactor-FXIOS-10657-connectETPSettingsNavigationWithBrowserVC&breakdown=afarcasanu%2Ffxios_10772_links_are_missing_roles&breakdown=afarcasanu%2Ffxios_10901_update_strings_for_data_preferences&breakdown=automation%2Fupdate-nimbus-experiments&breakdown=cc%2FFXIOS-10914_calculate-top-sites-asynchronously&breakdown=cc%2FFXIOS-FXIOS-10913_update-logic-for-top-sites-layout&test_case_name=testDispatchMultipleActions_mixThread%28%29&module=StoreTests&dateFrom=1704067200&dateTo=1735257599)